src/server/session.ts

```diff
@@ -120,6 +120,115 @@ namespace ts.server {
         export const Unknown = "unknown";
     }
 
+    module Metrics {
+        var eventCounts: Map<number> = {};
+        var properties: Map<string> = {};
+        let dtsCount = 0;
+
+        const sendInterval = 1000 * 60 * 5; // 5 minutes
+        let nextSendTimeMs = Date.now() + sendInterval;
+
+        const settingNames = [
+            "allowNonTsExtensions",
+            "declaration",
+            "emitBOM",
+            "emitDecoratorMetadata",
+            "experimentalAsyncFunctions",
+            "jsx",
+            "isolatedModules",
+            "module",
+            "moduleResolution",
+            "noEmit",
+            "noEmitOnError",
+            "noImplicitAny",
+            "noLib",
+            "noResolve",
+            "out",
+            "outFile",           
+            "preserveConstEnums",
+            "removeComments",
+            "suppressImplicitAnyIndexErrors",
+            "target"
+        ];
+
+        // Keeps a count of total invocations of various language service operations (rename, gotodef, etc)
+        export function countEvent(eventName: string, projectSvc: ProjectService, host: ts.System) {
+            let opts = projectSvc.getFormatCodeOptions();
+            if (opts.SendMetrics) {
+                eventCounts[eventName] = (eventCounts[eventName] || 0) + 1;
+
+                if (Date.now() > nextSendTimeMs) {
+                    registerSettings(projectSvc);                    
+                    send(host, opts.TelemetryUserID);
+                }
+            }
+        }
+
+        // Keeps track of any .d.ts files used which correspond to known versions from DefinitelyTyped
+        export function countDts(file: string) {
+            for(var i in properties) {

```

Comment:
@DanielRosenwasser - just out of curiosity, why do you not enforce that code in the TypeScript code based is formatted using [`tsfmt`](https://github.com/vvakame/typescript-formatter)?